typescript-selenium/tests/HappyPath.spec.ts

```diff
@@ -0,0 +1,165 @@
+/*********************************************************************
+ * Copyright (c) 2019 Red Hat, Inc.
+ *
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ **********************************************************************/
+
+import { e2eContainer } from "../inversify.config";
+import { Driver } from "../driver/Driver";
+import { TYPES, CLASSES } from "../inversify.types";
+import { DriverHelper } from "../utils/DriverHelper";
+import { By, WebElementCondition, Condition } from "selenium-webdriver";
+import { describe, after, test } from "mocha";
+import { LoginPage } from "../pageobjects/login/LoginPage";
+import { Dashboard } from "../pageobjects/dashboard/Dashboard";
+import { expect, assert } from 'chai'
+import { Workspaces } from "../pageobjects/dashboard/Workspaces";
+import { NameGenerator } from "../utils/NameGenerator";
+import { NewWorkspace } from "../pageobjects/dashboard/NewWorkspace";
+import { WorkspaceDetails } from "../pageobjects/dashboard/workspace-details/WorkspaceDetails";
+import { WorkspaceDetailsPlugins } from "../pageobjects/dashboard/workspace-details/WorkspaceDetailsPlugins";
+import { Request, post, get } from "selenium-webdriver/http";
+import { TestWorkspaceUtil } from "../utils/workspace/TestWorkspaceUtil";
+import { Ide } from "../pageobjects/ide/Ide";
+import { ProjectTree } from "../pageobjects/ide/ProjectTree";
+import { Editor } from "../pageobjects/ide/Editor";
+import { TestConstants } from "../TestConstants";
+import * as mocha from 'mocha'
+
+const workspaceName: string = NameGenerator.generate("wksp-test-", 5);
+const namespace: string = "che";
+const sampleName: string = "console-java-simple";
+
+const driver: Driver = e2eContainer.get<Driver>(TYPES.Driver);
+const driverHelper: DriverHelper = e2eContainer.get(CLASSES.DriverHelper);
+const loginPage: LoginPage = e2eContainer.get<LoginPage>(TYPES.LoginPage);
+const dashboard: Dashboard = e2eContainer.get(CLASSES.Dashboard);
+const workspaces: Workspaces = e2eContainer.get(CLASSES.Workspaces);
+const newWorkspace: NewWorkspace = e2eContainer.get(CLASSES.NewWorkspace);
+const workspaceDetails: WorkspaceDetails = e2eContainer.get(CLASSES.WorkspaceDetails);
+const workspaceDetailsPlugins: WorkspaceDetailsPlugins = e2eContainer.get(CLASSES.WorkspaceDetailsPlugins)
+const ide: Ide = e2eContainer.get(CLASSES.Ide)
+const projectTree: ProjectTree = e2eContainer.get(CLASSES.ProjectTree)
+const editor: Editor = e2eContainer.get(CLASSES.Editor)
+
+
+suite("E2E", async () => {
```

Comment:
These tests actually do not *test* anything - they just click and wait. I miss any kind of expectations - e.g. that the right labels are displayed or workspace has proper properties. I do not think that "waits" should replace "expectations".

Reply:
It waits for all things which this test should check. Immediate checks don't work properly, because we should wait condition before check. And if condition already checked, what sense to check it again with "expect".

Reply:
I do understand the intention but I do not agree with it. I think that test should only contain actions and expectations and the waiting for the data to load should be encapsulated into the page objects itself. For example the following snippet makes test easy understandable - user logs in and the dashboard is preset. The details of what does it mean and what should we wait for should be hidden into page object. In that way you can use it in different tests. 
```
   await loginPage.login()

   expect(await dashboard.isDisplayed()).to.be.true;
```

Reply:
It just about habit, for our team current state is more clear
```
 await loginPage.login()
 await dashboard.waitPage()
```

Reply:
I think it is very controversial question which approach more preferable with `waits` or` asserts`.
I've saw 3 approaches. Some guys use solution with asserts (for more understanding what we check), some companies uses just waits, someone use mixed style. For instance we tried use asserts in ***CHE6*** and we got a lot of side effects(for instanse: `assertEqual(element.getText(),'it's good'))`. It  be ok  if element won't be changed and we totally sure that element does not change statement. But if we have many elements which are changed dynamically - we can get for instance `StaleElementExeptions` in many asserts. Or we should check just short - term appearance  text and next moment - text disappear it also can be problem.
I think  choosing of approach dependents on individually for each application / project.
See also:
https://sqa.stackexchange.com/questions/7502/using-webdriverwait-to-verify-a-performed-action-versus-assertions-in-junit
https://imalittletester.com/2018/10/04/use-waits-as-assertions-for-your-selenium-tests/

Reply:
Maybe I have another point of view to this. We will be reusing the code in downstream. As from what I know now, I'm not able to rewrite just one ```test``` in a ```suite``` you wrote. That means, that the more logic you write to the ```test``` part and to the suites in general, the greater is the probability that we won't be able to use this suite and we would be forced to create a new one. If the logic is placed to the page object we can simply rewrite it. 
E.g. in ```wait dashboard``` test our are waiting for loader, loader disapperance and page. Now I'm not able to add any wait here. If that would be in page object, I can extend this to wait for some rh-che specific pop-ups for example.
Speaking about that, I would again rise an idea about having test layer with the less logic, page object with specification of the object itself and then some interaction layer that will deal with the page object interactions and behaviour, waits etc. I see benefit there to have these separated as it is easier to understand where to go if you want something to be changed. One page object with all of its definitions and interactions could be quite large and hard to read.

Reply:
I agree that each team should choose the approach that suits it best, but since you're starting these TypeScript UI tests from scratch I thought this is the best time to discuss another possibilities. 

It is not only about asserts vs. waits. It is also about what should the Page Object do. For me, it should not be just another representation of DOM. It should represent the page as user sees it with data on page and available actions. It should also use domain language like "project" instead of "item". So I like your LoginPage with only one method - login. It is neat and it is all that user needs from Login page. 

However, the ProjectTree  contains methods like collapse item, expand item which I find too low level for e2e tests. I think that something like selectProject or getProject would be more suitable. There are also 7 different wait methods. But why does the client of this class need to call wait methods at all? Those should be part of every action automatically. Like when I want to openFile I want automaticaly to wait until it is open. 

You can see this approach in Red Deer project. It is a framework for testing Eclipse.

https://github.com/eclipse/reddeer/blob/37f27affd709a9622d84e8d29d047dc78a0de1b4/plugins/org.eclipse.reddeer.eclipse/src/org/eclipse/reddeer/eclipse/ui/navigator/resources/AbstractExplorer.java

Unlike Red Deer, which is a general framework for testing Eclipse, I'd suggest that you implement only those methods that you actually need. So I would delete methods like collapseItem because when they are not used they will start to not function over time. And it is harder to debug functions that "should" work than to write them from scratch. 

As I've written at the beginning, my intention is not to criticise the work you've done. The PR looks good. I just wanted to exchange some experience that we\ve gathered in our team over the past couple of years.

Reply:
I agree with Katka - that would clearly define the responsibilities of different classes. Page object for interaction with page elements, test for defining what should be done and the "interaction" layer between them (logging, error handling, waits).

Reply:
So, about reusing code and representation in PageObjecst +1. @Katka92, Lucia Jelinkova thanks for feedback.

Reply:
I guess we have a really great experience exchange. Thank you for so big engagement and precision review.